### PR TITLE
Encrypt configuration cache unconditionally

### PR DESCRIPTION
**Pull request by bamboo**
- Replace env vars with keystore as source of encryption keys.
- Unconditionally encrypt the CC state files

Fixes: https://github.com/gradle/gradle/issues/23866

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC

**Comment by bamboo** [comment]
@bot-gradle test APT

**Comment by bamboo** [comment]
@bot-gradle test and merge

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy [hunk 1]
```diff
@@ -162,110 +137,89 @@ class ConfigurationCacheEncryptionIntegrationTest extends AbstractConfigurationC
...
-        runWithEncryption(false)
+        runWithEncryption()
+        findKeystoreFile().delete()
 
         when:
-        runWithEncryption(true)
+        runWithEncryption()
 
         then:
         configurationCache.assertStateStored()
```
Comment (abstratt) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy:150]: Could add a check here to ensure the keystore was recreated at original location (assumin that is what is expected)

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt [hunk 2]
```diff
@@ -57,49 +61,67 @@ interface EncryptionService : EncryptionConfiguration {
...
+
+            else -> {
                 null
             }
+        }
     }
 
+    private
+    fun secretKeyFrom(keySource: KeyStoreKeySource, cacheBuilderFactory: GlobalScopedCacheBuilderFactory) =
+        // JKS does not support non-PrivateKeys
```
Comment (abstratt) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt:103]: This comment could be near `loadOrCreateKey` or the `KEYSTORE_TYPE` constant.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/FileLockManager.java [hunk 3]
```diff
@@ -16,10 +16,13 @@
 package org.gradle.cache;
 
 import org.gradle.api.Action;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
 
 import javax.annotation.Nullable;
 import java.io.File;
 
+@ServiceScope(Scope.Global.class)
```
Comment (abstratt) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/FileLockManager.java:25]: Still want this change?

**Review by abstratt — APPROVED** [review]
LGTM, thanks for taking this over!

Left mostly questions and feedback that is not critical do be addressed

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt [hunk 4]
```diff
@@ -179,59 +191,54 @@ class KeyStoreKeySource(
     private
     val keyProtection = KeyStore.PasswordProtection(CharArray(0))
 
+    private
+    val keyStore by lazy {
```
Comment (abstratt) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt:195]: note keystore is not supposed to be thread-safe. Are we guaranteeing somehow that at most one thread ever interacts with a keystore instance?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/FileLockManager.java (hunk 3 above)
Comment (bamboo) [reply subprojects/persistent-cache/src/main/java/org/gradle/cache/FileLockManager.java:25]: Yes. It's leaving the campsite better.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt (hunk 2 above)
Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt]: Done, thanks.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt (hunk 4 above)
Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt:199]: `KeyStoreKeySource` instances are not shared between threads, the produced `SecretKey` instance might be shared but (without checking) I wouldn't expect that to be a problem.

**Comment by bamboo** [comment]
Thanks for the review, @abstratt!

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt [hunk 5]
```diff
@@ -57,49 +61,83 @@ interface EncryptionService : EncryptionConfiguration {
...
+    fun secretKeyFrom(keySource: KeyStoreKeySource, cacheBuilderFactory: GlobalScopedCacheBuilderFactory) =
+        cacheBuilderFactory
+            .run { keySource.customKeyStoreDir?.let { createCacheBuilderFactory(it) } ?: this }
+            .createCacheBuilder("cc-keystore")
+            .withDisplayName("Gradle Configuration Cache keystore")
+            .withInitializer {
+                keySource.createKeyStoreAndGenerateKey(keyStoreFile())
+            }
+            .open().useToRun {
+                try {
```
Comment (abstratt) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt:116]: both `assert keystoreFile().file` and `assert keystoreFile().length() > 0` should always pass here, right?

Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt:116]: Right, unless the cache has been tampered with.

**Comment by bamboo** [comment]
@bot-gradle test APT

**Comment by bamboo** [comment]
@bot-gradle test and merge

**Pull request merged**